pkg/s3/bucket.go

```diff
@@ -179,44 +183,73 @@ func TagBucket(s3Client *s3.S3, bucketName string, backUpLocation string, infraN
 	return nil
 }
 
-// FindExistingBucket looks for an S3 bucket matching this cluster's velero tags
-// and infrastructure tags. If a matching bucket is found, the bucket name is returned.
-func FindExistingBucket(s3Client *s3.S3, infraName string) (string, error) {
-	// List all buckets associated with this cluster's AWS account.
+// ListBuckets lists all buckets in the AWS account.
+func ListBuckets(s3Client Client) (*s3.ListBucketsOutput, error) {
 	input := &s3.ListBucketsInput{}
 	result, err := s3Client.ListBuckets(input)
 	if err != nil {
 		fmt.Println(err.Error())
-		return "", err
+		return result, err
 	}
+	return result, nil
+}
 
-	for _, bucket := range result.Buckets {
-		request := &s3.GetBucketTaggingInput{
-			Bucket: aws.String(*bucket.Name),
+// ListBucketTags returns a list of s3.GetBucketTagging objects, one for each bucket.
+// If the bucket is not readable, or has no tags, an empty TagSet is returned for that bucket.
+func ListBucketTags(s3Client Client, bucketlist *s3.ListBucketsOutput) (map[string]*s3.GetBucketTaggingOutput, error) {
+	taglist := make(map[string]*s3.GetBucketTaggingOutput)
+	emptyTagSet := &s3.GetBucketTaggingOutput{
+		TagSet: []*s3.Tag{},
+	}
+	for _, bucket := range bucketlist.Buckets {
+		taglist[*bucket.Name] = emptyTagSet
+		// Sometimes deleted buckets will show up in this list.
+		// In case they are in the process of being deleted, exit gracefully.
+		bucketReadable, err := DoesBucketExist(s3Client, *bucket.Name)
```

Comment:
Is there a reason for doing this check, as opposed to handling the case where the `ec2err.Code() == "NoSuchBucket"`? This seems on the surface like an unnecessary API call.

Reply:
Basically I was avoiding copy/pasting the logic that handles the NoSuchBucket error. Since it's already implemented in the DoesBucketExist function, I figured that calling the function would be an effective way to determine if the bucket is in a readable state.

Though now that you bring it up, given that the API can sometimes say "everything is ok" and other times say "nope, not readable", I think any amount of one-time check might not be adequate. So my second API call is a bad idea in this case.

Maybe instead of a separate API call, I should implement the error-checking logic in a specific error-checking function, and call that function anywhere that a flaky API response could cause an issue.

I'll look into doing this tomorrow. Thanks for the suggestion!

Reply:
You're already evaluating ec2err.Code here, so perhaps even just a simple switch on what that error code is would suffice. In this case, simple and readable can be preferred over "Don't Repeat Yourself".. and especially true in this case where we are actually making network calls to the API.

Reply:
That sounds great. I'll give it a try.